File: src/main/java/com/github/jferard/classwriter/api/MethodBuilder.java
/*
 * ClassWriter - A minimal Java bytecode writer. Creates classes, methods, interfaces...
 *     Copyright (C) 2018 J. Férard <https://github.com/jferard>
 *
 * This file is part of ClassWriter.
 *
 * ClassWriter is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * ClassWriter is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.jferard.classwriter.api;

import com.github.jferard.classwriter.TypedIdentifier;
import com.github.jferard.classwriter.internal.attribute.Annotation;
import com.github.jferard.classwriter.internal.attribute.CodeAttribute;
import com.github.jferard.classwriter.internal.attribute.stackmap.VerificationType;
import com.github.jferard.classwriter.internal.instruction.Instruction;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * 4.6. Methods
 */
public class MethodBuilder {
    private final CodeAttribute codeAttribute;
    private final MethodDescriptor descriptor;
    private final String name;
//    private List<Writable> attributes;
    private int accessFlags;
    private List<TypedIdentifier> params;
    private TypedIdentifier ret;
    private List<Exception> exceptions;
    private boolean synthetic;
    private boolean deprecated;
    private String signature;
    private List<Annotation> runtimeVisibleParameterAnnotations;
    private List<Annotation> runtimeInvisibleParameterAnnotations;
    private List<Annotation> runtimeVisibleAnnotations;
    private List<Annotation> runtimeInvisibleAnnotations;
    private List<Annotation> runtimeVisibleTypeAnnotations;
    private List<Annotation> runtimeInvisibleTypeAnnotations;
    private Object annotationDefault;
    private Object methodParameters;

    public MethodBuilder(final String name, final MethodDescriptor descriptor, Instruction code) {
        this.name = name;
        this.descriptor = descriptor;
//        this.attributes = new ArrayList<>();
        final List<VerificationType> argVerificationTypes = descriptor.getArgTypes().stream()
                .map(at -> at.getVerificationType()).collect(Collectors.toList());
        this.codeAttribute = new CodeAttribute(code, argVerificationTypes, new ArrayList<>());
    }

    public MethodBuilder accessFlags(final int accessFlags) {
        this.accessFlags = accessFlags;
        return this;
    }

    public MethodBuilder params(final List<TypedIdentifier> params) {
        this.params = params;
        return this;
    }

    public MethodBuilder ret(final TypedIdentifier ret) {
        this.ret = ret;
        return this;
    }

    public MethodBuilder exceptions(final List<Exception> exceptions) {
        this.exceptions = exceptions;
        return this;
    }

    public MethodBuilder synthetic() {
        this.synthetic = true;
        return this;
    }

    public MethodBuilder deprecated() {
        this.deprecated = true;
        return this;
    }

    public MethodBuilder signature(final String signature) {
        this.signature = signature;
        return this;
    }

    public MethodBuilder addRuntimeVisibleParameterAnnotation(final Annotation annotation) {
        this.runtimeVisibleParameterAnnotations.add(annotation);
        return this;
    }

    public MethodBuilder addRuntimeInvisibleParameterAnnotation(final Annotation annotation) {
        this.runtimeInvisibleParameterAnnotations.add(annotation);
        return this;
    }

    public MethodBuilder addRuntimeVisibleAnnotation(final Annotation annotation) {
        this.runtimeVisibleAnnotations.add(annotation);
        return this;
    }

    public MethodBuilder addRuntimeInvisibleAnnotation(final Annotation annotation) {
        this.runtimeInvisibleAnnotations.add(annotation);
        return this;
    }

    public MethodBuilder addRuntimeVisibleTypeAnnotation(final Annotation annotation) {
        this.runtimeVisibleTypeAnnotations.add(annotation);
        return this;
    }

    public MethodBuilder addRuntimeInvisibleTypeAnnotation(final Annotation annotation) {
        this.runtimeInvisibleTypeAnnotations.add(annotation);
        return this;
    }

    public MethodBuilder annotationDefault(final Annotation AnnotationDefault) {
        this.annotationDefault = AnnotationDefault;
        return this;
    }

    public MethodBuilder methodParameters(Object methodParameters) {
        this.methodParameters = methodParameters;
        return this;
    }

    public MethodBuilder access(int accessFlags) {
        this.accessFlags = accessFlags;
        return this;
    }

    public Method build() {
        return new Method(this.accessFlags, this.name, this.descriptor, this.codeAttribute);
    }
}

File: src/main/kotlin/com/github/jferard/classwriter/api/MethodBuilder.kt
/*
 * ClassWriter - A minimal JVM bytecode writer. Creates classes, methods, interfaces...
 *     Copyright (C) 2018 J. Férard <https://github.com/jferard>
 *
 * This file is part of ClassWriter.
 *
 * ClassWriter is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * ClassWriter is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.jferard.classwriter.api

import com.github.jferard.classwriter.TypedIdentifier
import com.github.jferard.classwriter.internal.attribute.Annotation
import com.github.jferard.classwriter.internal.attribute.CodeAttribute
import com.github.jferard.classwriter.api.instruction.Instruction
import java.util.*

/**
 * 4.6. Methods
 */
class MethodBuilder(private val name: String,
                    private val descriptor: MethodDescriptor,
                    code: Instruction) {
    private val codeAttribute: CodeAttribute
    //    private List<Writable> attributes;
    private var accessFlags = 0
    private var params: MutableList<TypedIdentifier>? = null
    private var ret: TypedIdentifier? = null
    private var exceptions: MutableList<Exception>? = null
    private var synthetic = false
    private var deprecated = false
    private var signature: String? = null
    private val runtimeVisibleParameterAnnotations: MutableList<Annotation> = ArrayList()
    private val runtimeInvisibleParameterAnnotations: MutableList<Annotation>? = null
    private val runtimeVisibleAnnotations: MutableList<Annotation>? = null
    private val runtimeInvisibleAnnotations: MutableList<Annotation>? = null
    private val runtimeVisibleTypeAnnotations: MutableList<Annotation>? = null
    private val runtimeInvisibleTypeAnnotations: MutableList<Annotation>? = null
    private var annotationDefault: Any? = null
    private var methodParameters: Any? = null

    fun accessFlags(accessFlags: Int): MethodBuilder {
        this.accessFlags = accessFlags
        return this
    }

    fun params(params: MutableList<TypedIdentifier>): MethodBuilder {
        this.params = params
        return this
    }

    fun ret(ret: TypedIdentifier): MethodBuilder {
        this.ret = ret
        return this
    }

    fun exceptions(exceptions: MutableList<Exception>): MethodBuilder {
        this.exceptions = exceptions
        return this
    }

    fun synthetic(): MethodBuilder {
        synthetic = true
        return this
    }

    fun deprecated(): MethodBuilder {
        deprecated = true
        return this
    }

    fun signature(signature: String): MethodBuilder {
        this.signature = signature
        return this
    }

    fun addRuntimeVisibleParameterAnnotation(
            annotation: Annotation): MethodBuilder {
        runtimeVisibleParameterAnnotations!!.add(annotation)
        return this
    }

    fun addRuntimeInvisibleParameterAnnotation(
            annotation: Annotation): MethodBuilder {
        runtimeInvisibleParameterAnnotations!!.add(annotation)
        return this
    }

    fun addRuntimeVisibleAnnotation(
            annotation: Annotation): MethodBuilder {
        runtimeVisibleAnnotations!!.add(annotation)
        return this
    }

    fun addRuntimeInvisibleAnnotation(
            annotation: Annotation): MethodBuilder {
        runtimeInvisibleAnnotations!!.add(annotation)
        return this
    }

    fun addRuntimeVisibleTypeAnnotation(
            annotation: Annotation): MethodBuilder {
        runtimeVisibleTypeAnnotations!!.add(annotation)
        return this
    }

    fun addRuntimeInvisibleTypeAnnotation(
            annotation: Annotation): MethodBuilder {
        runtimeInvisibleTypeAnnotations!!.add(annotation)
        return this
    }

    fun annotationDefault(
            AnnotationDefault: Annotation): MethodBuilder {
        annotationDefault = AnnotationDefault
        return this
    }

    fun methodParameters(methodParameters: Any): MethodBuilder {
        this.methodParameters = methodParameters
        return this
    }

    fun access(accessFlags: Int): MethodBuilder {
        this.accessFlags = accessFlags
        return this
    }

    fun build(): Method {
        return Method(accessFlags, name,
                descriptor, codeAttribute)
    }

    init {
        //        this.attributes = new ArrayList<>();
        val argVerificationTypes =
                descriptor.argTypes
                        .map { at: ValueType -> at.verificationType }
        codeAttribute = CodeAttribute(code, argVerificationTypes,
                ArrayList())
    }
}